docs/proposals/deployment.md

```diff
@@ -0,0 +1,172 @@
+a ReplicationController to bring up the desired pods.
+Users can also target the Deployment at existing ReplicationControllers, in
+which case the new RC will replace the existing ones. The exact mechanics of
+replacement depends on the DeploymentStrategy chosen by the user.
+DeploymentStrategies are explained in detail in a later section.
+
+## Implementation
+
+### API Object
+
+The `Deployment` API object will have the following structure:
+
+```go
+type Deployment struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior of the Deployment.
+  Spec DeploymentSpec
+
+  // Most recently observed status of the Deployment.
+  Status DeploymentStatus
+}
+
+type DeploymentSpec struct {
+  // Number of desired pods. This is a pointer to distinguish between explicit
+  // zero and not specified. Defaults to 1.
+  Replicas *int
+
+  // Label selector for pods. Existing ReplicationControllers which have all these
+  // label selectors will be scaled down.
+  Selector map[string][string]
+
+  // Describes the pods that will be created.
+  Template *PodTemplateSpec
+
+  // The deployment strategy to use to replace existing pods with new ones.
+  Strategy DeploymentStrategy
+}
+
+type DeploymentStrategy struct {
+  // Type of deployment. Can be "Recreate" or "RollingUpdate".
+  Type DeploymentType
+
+  // Rolling update config params. Present only if DeploymentType =
+  // RollingUpdate.
+  RollingUpdate *RollingUpdateDeploymentSpec
+}
+
+type DeploymentType string
+
+const (
+  // Kill all existing pods before creating new ones.
+  DeploymentRecreate DeploymentType = "Recreate"
+
+  // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+  DeploymentRollingUpdate DeploymentType = "RollingUpdate"
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
+  // The percentage of pod replicas that should be updated in each interval.
+  // If nil, a fixed value of 1 pod is used.
+  UpdatePercent *int
+
+  // While updating pods, a "readiness" check is done instead of just the
+  // pod "creation" check, if this is true. False, by default.
+  UseReadinessAsAcceptance bool
+}
+
+type DeploymentStatus struct {
+  // The total number of pods created with the desired template spec.
+  Replicas int
+
+  // The most recent generation observed by the controller.
+  ObservedGeneration int64
+}
+
+```
+
+### Controller
+
+The DeploymentController will make a given Deployment happen.
+Whenever a new Deployment is created, it will:
+
+1. Find all RCs with the same label selector as DeploymentSpec.Selector.
+2. Add annotations for bookkeeping:
+   - Add "deployment.kubernetes.io/<deployment-id>" annotation to the existing
+     RCs to identify the deployment controller that is controlling them.
+3. Create a new RC to create pods using DeploymentSpec.PodTemplate. Set replicas = 0.
```

Comment:
No the new RC will have replicas = 0 to begin with.
It will be scaled up as per the deploymentStrategy. We might want to scale the existing RC down first before scaling up this new RC (to preserve resources).